### ISSUE TITLE
[Bug Report] Can't cast any video to Chromecast

### ISSUE THREAD

**Issue by hecklawert** [stashapp/stash#4136]
**Describe the bug**
Every time I try to cast any video to Chromecast, the player hangs and doesn't play the video. Chromecast doesn't show any error; it only displays a cast icon. I've tried this with both HTTP and HTTPS, and I get the same error.

**To Reproduce**
1. Go to any video
2. Click on ChromeCast Button
4. Player hangs and doesn't play the video

**Expected behavior**
Video should play at TV with Chromecast

**Stash Version: (from Settings -> About):**
v0.22.1

**Desktop:**
 - OS: Linux Fedora 34 and Linux Fedora 38
 - Browser: Chrome
 - Version: 117.0.5938.62 Stable

**Smartphone:**
 - Device: One Plus Nord
 - OS: Android 10
 - Browser: Chrome and Brave

**Comment by Eviepayne** [comment]
I also encounter this.
Not sure why it happens.
I have https, stash works normally.
the chromecast worked once and then never again.

**Comment by Eviepayne** [comment]
@hecklawert I resolved my issue.
Do you have your stash on reverse proxy? I found that the issue was DNS.
chromecast hardcodes 8.8.8.8 as it's DNS provider. I added a outbound NAT on my firewall to redirect 8.8.8.8 to my local dns and it resolved my issue (note that my local dns had my stash domain)

tl;dr it could be you need to either do what I did or make your domain public DNS.

**Comment by Qlupic** [comment]
I have the same problem and I tried to change DNS to Google both in the router and windows. Didn't help. I am using Windows 10. YouTube casting works as normal.

**Comment by Eviepayne** [comment]
> I have the same problem and I tried to change DNS to Google both in the router and windows. Didn't help. I am using Windows 10. YouTube casting works as normal.

The issue isnt what dns you use. Its that google doesnt know your hostname assuming you have a domain on it.

you have 2 options.

(Not recommended)

1. publish the dns on the internet and make stash public facing.

(Recommended)

2. In your router create a NAT rule that redirects requests from google to your own dns resolver.

For ex. in my opnsense firewall I redirect all DNS queries to my Opnsense (localhost)
The DNS there is a recursor so I get internet published dns as well.

in your case you may want to only redirect 8.8.8.8 unless your dns server is a recursor as well.

**Comment by Kylindrias** [comment]
I have the same issue. I'm not using a reverse proxy, server is on the same LAN. Tried using both the local DNS entry and the IP address but both just display a cast icon on the TV and then nothing happens.

Is there an issue caused by trying to cast from the port 9999?

**Comment by Eviepayne** [comment]
I might wrong about this but I think cast requires a working TLS cert.

**Comment by Eviepayne** [comment]
Nope needs to be valid

**Comment by Qlupic** [comment]
I managed to workaround this issue by using Chrome browser on Android as it seems none of the other browsers would work. This worked even when the Stash docker is behind a VPN, but with LAN access enabled.

**Comment by garfolino** [comment]
@DogmaDragon I use chrome cast a looooot, and not having this working on stash is a shame. Can I take on it?

Here are the [details](https://github.com/dude-stash/stash/commit/5362acc26dfbd1470c78b3c32d861b37621a7c10) of the issue, and proposed plan. 

I already worked on a poc on the weekend, and it works. Now im rewriting from scratch in a more organized fashion.

FYI, I tried the plugin route, but it requires changes on the app anyway, so a plugin alone couldn't solve it

[VIDEO HERE](https://www.loom.com/share/52410e4b8dae4f7982dee251443bd8a6)

**Comment by Eviepayne** [comment]
@garfolino Claude is most likely wrong here.
I can use cast fine. The most likely issue you're experiencing is what I detailed above.

**Comment by garfolino** [comment]
> [@garfolino](https://github.com/garfolino) Claude is most likely wrong here. I can use cast fine. The most likely issue you're experiencing is what I detailed above.

Can you share your setup? and more details? I looked over GitHub, discourse and discord, and the majority of people has issues with it. The explanation from the research I did is solid and the changes I made makes it work seamlessly. 

The only think I can do on Chrome is cast the tab itself, which is not as efficient as casting the actual video. But id be intrigued to hear about your experience.

Edit: I saw your comments earlier. You had to do a workaround that not everyone is able to do, or should do. A proper implementation makes it work for everyone out of the box

**Comment by Eviepayne** [comment]
Most people aren't sysadmins so they don't know what's wrong but Chromecast is basically an embedded device with a chrome browser in kiosk mode. It also hard-codes its DNS to Google's 8.8.8.8

So first you need a valid TLS cert, which means buying a domain, configuring let's encrypt or another ACME provider. Next you need to intercept the DNS request in your router and redirect it to a DNS server you control and have configured to point to the correct server.

I use opnsense as my firewall, and unbound for DNS. I own a domain and configured haproxy to serve my stash.

From there I configured a DNAT rule to redirect port 53 traffic (DNS) to my local unbound where I made a custom record for my stash. I have also configured an acme client to automatically renew and deploy TLS certs into my haproxy.

Its a lot of configuring but thats what happens when you use Google products. You have to manipulate them to work.

**Comment by garfolino** [comment]
> Most people aren't sysadmins so they don't know what's wrong but Chromecast is basically an embedded device with a chrome browser in kiosk mode. It also hard-codes its DNS to Google's 8.8.8.8

Exactly. And that's why apps like Stash should have functionalities that work out of the box. Also, if you were running your stash on a laptop and took it on a field trip, your chrome cast wouldn't work outside your well configured LAN. 

The changes here are so anyone, from day 1 can click a button and cast their movies. Similar to Plex. Similar to Pornhub, or YouTube.

**Comment by garfolino** [comment]
> [@DogmaDragon](https://github.com/DogmaDragon) I use chrome cast a looooot, and not having this working on stash is a shame. Can I take on it?
> 
> Here are the [details](https://github.com/dude-stash/stash/commit/5362acc26dfbd1470c78b3c32d861b37621a7c10) of the issue, and proposed plan.
> 
> I already worked on a poc on the weekend, and it works. Now im rewriting from scratch in a more organized fashion.
> 
> FYI, I tried the plugin route, but it requires changes on the app anyway, so a plugin alone couldn't solve it
> 
> [VIDEO HERE](https://www.loom.com/share/52410e4b8dae4f7982dee251443bd8a6)

@DogmaDragon I understand its quite a big PR, and Im happy to break it down, and clean up a bit. Whatever you prefer. I have plans for a second iteration, so the internal player controls the media on tv as well, but I kept that out of this PR for now. I also didn't open a PR against stash yet, only to my own fork, for reference.
